### ISSUE TITLE
Add feature to provide performance profiles

### ISSUE THREAD

**Issue by ktenzer** [alexandrevilain/temporal-operator#129]
This likely won't be an exact science as performance can vary drastically depending on workload. However there are likely some patterns we can extrapolate and some knobs / dials that would work well with those assumptions.

We have a sliding multiplier (we have a base and it can be increased by 1-5x for example). Ideally we would do STNs (state transitions per second) but I don't think this would be possible as there is too much variance.

We would tune things like shards and other parameters as well as how many replicas of each type (frontend, history, matching, worker) and of course their resource limits (CPU/Mem).

We could also deploy our benchmark tool maru: https://github.com/temporalio/maru as a way to allow users to verify easily.

For the spec I was thinking something like
<pre>
...
spec:
  performance:
    profile: <base, 1x,2x, 3x, 4x, 5x>     
... 
</pre>

**Comment by alexandrevilain** [comment]
Hi!

Great idea! Maybe we can make a draft table to determine profiles names and their default values  ?  
For instance:

| Profile name | Shards count | Frontend replicas counts | History replicas count | Matching replicas counts|
| --- | --- | --- | --- |--- |
| dev | 1 | 1 | 1 | 1 | 
| base | 9 | 3 | 3 | 3 | 

What do you think about that ? IMO it would help us before starting any coding.

**Comment by ktenzer** [comment]
Linking to current server docs which go into some of this topic: https://github.com/temporalio/documentation/blob/master/docs/server/production-deploy.md

**Comment by ktenzer** [comment]
> 



> Hi!
> 
> Great idea! Maybe we can make a draft table to determine profiles names and their default values ? For instance:
> 
> Profile name	Shards count	Frontend replicas counts	History replicas count	Matching replicas counts
> dev	1	1	1	1
> base	9	3	3	3
> What do you think about that ? IMO it would help us before starting any coding.

yep I was thinking some kind of table to capture the knobs/dials. Let me do some more research on that and then I will iterate on this table.

**Comment by ktenzer** [comment]
Profile name | Profile Type | Autoscale | Pod Size | Scale Slider %
-- | -- | -- | -- 
dev | baseline | disabled | small | n/a
test | baseline | enabled | medium| n/a
prod | custom | disabled | large | 500

We would maintain a replia count for frontend:history:matching:worker of 1:2:1:1. So baseline would go with that. Choosing customer allows you to apply scale slider in increments of 100%. So setting slider to 100 would have replica counts of 2:4:2:2.

Shards we would keep 512. As you add more pods via slider you get more shards and that is good starting point.

Pod size would be tiny, small, medium, large. We would set hard limits at below value and request half.
-- tiny 1 core 1GB RAM
-- small 4 core 4GB RAM
-- medium 6 core 6GB RAM
-- large 8 core 8GB RAM

@alexandrevilain WDYT?

**Comment by alexandrevilain** [comment]
Hi! 

LGTM 👍

The main issue I see for this features, is the complexity. 

IMO, A new user would love the `dev` or `test` profiles. It's a good and easy starting point to start with temporal on Kubernetes.

But the SRE wanting to go to production using Temporal, would certainly customize all the configurations to suits his needs.

So, does this whole complex solution worth it ? 

Shouldn't we focus on the operator stability and ease of deploying before starting to develop this kind of features without any user feedback ? Maybe you already have some of them ?

**Comment by ktenzer** [comment]
that is good point and I agree we should get some more user feedback. The initial intention was apply some good practice while making it easy to deploy. Maybe for starters we just apply the 1:2:1:1 pod ratio, 512 shards and resource limit t-shirt sizes? We can then later implement or think about profiles, autoscaling, etc?

**Comment by alexandrevilain** [comment]
>  Maybe for starters we just apply the 1:2:1:1

You mean, for now creating few profiles with only pod ratio, shards an ressouces limits ?
If that true, it seems fine and we can go for it now ! 👍

> limit t-shirt sizes

My English is not so good, I don't get it. What do you mean by that ? 

> We can then later implement or think about profiles, autoscaling, etc? 

For sure, even after implementing "basic/starters" profiles, we can keep this issue open to gather some feedbacks.

**Comment by ktenzer** [comment]
yes so my initial proposal is as follows:
- Deploy 1:2:1:1 pod ratio (so history is 2 pods)
- Set shards to 512
- Expose pod resource limits and set default to what I have listed as tiny
